### PR TITLE
[KernelGen][Nvidia] Add special_bessel_j1 operator with Triton kernel

### PR DESCRIPTION
**Pull request by XDYuanzhuLee**
## Summary
Adds a Triton kernel for `special_bessel_j1`. Computes the first-kind Bessel function of order 1 for each element of input.

## Testing
- Parametrized tests over SPECIAL_SHAPES and FLOAT_DTYPES
- Validated against reference on device via `to_reference(inp, True)`
- Tested on: Nvidia, Tianshu, Muxi, Ascend, Hygon

## Performance
Test command: `pytest benchmark/test_special_bessel_j1.py --level core` (NVIDIA H20)

| Configuration | Torch Latency (ms) | Gems Latency (ms) | Speedup | TFLOPS |
|---|---|---|---|---|
| [1073741824] | 49.753 | 6.467 | 7.694 | 0.166 |
| [64, 64] | 0.012 | 0.007 | 1.714 | 0.001 |
| [4096, 4096] | 0.795 | 0.109 | 7.321 | 0.155 |
| [64, 512, 512] | 0.794 | 0.109 | 7.311 | 0.155 |
| [1024, 1024, 1024] | 49.386 | 6.443 | 7.665 | 0.167 |
| **Arithmetic Mean** | — | — | **6.341** | — |

## Multi-backend Testing
| Backend | Accuracy Test | Benchmark | Speedup (mean) | Notes |
|---|---|---|---|---|
| Nvidia (H20) | PASS | PASS (5 cases, --level core) | 6.341 | Primary |
| Tianshu | FAIL | PASS (14 cases) | 1.276 | AssertionError: Scalars are not close |
| Muxi | FAIL | FAIL | — | CUDA error: memory size or pointer value too large |
| Ascend | FAIL | N/A | — | CompilationError |
| Hygon | PASS | PASS (14 cases) | 9.961 | — |

## Files Changed
- `src/flag_gems/ops/special_bessel_j1.py`: Triton kernel implementation
- `tests/test_special_bessel_j1.py`: Accuracy test
- `benchmark/test_special_bessel_j1.py`: Performance benchmark
- `src/flag_gems/ops/__init__.py`: Register import and `__all__`
- `src/flag_gems/__init__.py`: Register to `_FULL_CONFIG`
- `conf/operators.yaml`: Add operator entry (kind: Math, stage: alpha 5.1)

### CODE REVIEW

File: benchmark/test_special_bessel_j1.py [hunk 1]
```diff
@@ -0,0 +1,16 @@
+# Generated by KernelGen: https://github.com/flagos-ai/KernelGen
+import pytest
+import torch
+
+from . import base
+
+
+# bessel_j1_cuda does not support Half/BFloat16
```
Comment (tengqm) [inline benchmark/test_special_bessel_j1.py]: Since this operator is meant to be a generic one that will run on all backends, do we want to limit this operator to fp32 on all backends or treat NVIDIA as a special case?

File: tests/test_special_bessel_j1.py [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+import torch
+
+import flag_gems
+
+from . import accuracy_utils as utils
+
+
+@pytest.mark.special_bessel_j1
+@pytest.mark.parametrize("shape", utils.POINTWISE_SHAPES)
+@pytest.mark.parametrize("dtype", utils.FLOAT_DTYPES)
```
Comment (tengqm) [inline tests/test_special_bessel_j1.py:16]: This may lead to some confusions in the future ... why do we test all float types while only benchmark fp32?

File: benchmark/test_special_bessel_j1.py (hunk 1 above)
Comment (XDYuanzhuLee) [reply benchmark/test_special_bessel_j1.py]: This is a valid question. The short answer is: the fp32-only restriction is indeed NVIDIA-specific, not a Triton kernel limitation.

**Why fp32 only in benchmark:**
- The benchmark framework compares our Triton kernel against a reference  — . On CUDA, this dispatches to  which only has float32/double kernels via libdevice. Half/BFloat16 tensors would either error out or silently fall back to CPU, making the latency comparison meaningless.
- This is not a Triton kernel limitation: the kernel upcasts to  internally, so it handles fp16/bf16 correctly — as confirmed by the test.

**Should this be backend-specific rather than global?**
- In principle yes.  and  pass  (fp16, fp32, bf16) because  on CUDA supports all three.  can't because the CUDA reference doesn't. On non-NVIDIA backends, the reference might support broader dtypes.
- The current hardcoded  is a pragmatic safe default — it works on all backends. A per-backend dtype list would be the right long-term approach, but that requires framework changes to the benchmark harness, which is out of scope for this operator PR.

I'll add a more precise comment in the benchmark file clarifying this is an NVIDIA-specific limitation, not a kernel limitation.

File: tests/test_special_bessel_j1.py (hunk 2 above)
Comment (XDYuanzhuLee) [reply tests/test_special_bessel_j1.py:16]: Good observation — the asymmetry can indeed be confusing at a glance. Here is the reasoning:

**Test covers all  (fp16, fp32, bf16):** The purpose is to validate the kernel's **mathematical correctness** across all supported input types. The kernel internally upcasts to  before calling , then casts back on output — so the Bessel computation should be correct regardless of input dtype. Testing all of them ensures there's no dtype-casting bug in the pipeline.

**Benchmark only covers fp32:** The benchmark compares speed against  as a reference. On CUDA,  only supports float32/double — it has no Half/BFloat16 kernels. So benchmarking fp16/bf16 would produce a meaningless comparison (the reference would either error or CPU-fallback).

**Why  is different:**  on CUDA dispatches to  which supports all float types, so its benchmark can use . The bessel_j1 situation is specific to libdevice's  being fp32/double only.

I'll add a comment in the test file pointing to this explanation so future readers don't trip over the same question.

**Comment by w1120029931-bit** [comment]
please resolve conflicts

**Comment by XDYuanzhuLee** [comment]
Resolved merge conflicts and rebuilt branch on current upstream master. Added Apache 2.0 header to special_bessel_j1.py. 18/18 tests pass. Branch updated.

File: src/flag_gems/ops/special_bessel_j1.py [hunk 3]
```diff
@@ -0,0 +1,34 @@
+# Copyright 2026 Yuanzhu Li, Beijing Academy of Artificial Intelligence (BAAI)
+# and KernelGen (https://github.com/flagos-ai/KernelGen). All rights reserved.
+#
```
Comment (tengqm) [inline src/flag_gems/ops/special_bessel_j1.py]: Please don't change the copyright claim.
We intentionally claim copyrights for all "The FlagOS Contributors".

Comment (XDYuanzhuLee) [reply src/flag_gems/ops/special_bessel_j1.py]: Done.

**Comment by XDYuanzhuLee** [comment]
Resolved in the latest push. The conflict in `conf/operators.yaml` has been fixed by merging upstream/master. PR is now MERGEABLE. Validation: `git diff --check` clean, `py_compile` passes.

File: benchmark/test_special_bessel_j1.py [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+import torch
+
+from . import base
+
+
+# torch.special.bessel_j1 on CUDA dispatches to bessel_j1_cuda which only supports
+# float32/double. The Triton kernel handles all float dtypes (tested in test), but
+# we benchmark fp32 only because the CUDA reference can't run Half/BFloat16 on GPU.
+# On non-NVIDIA backends the reference may support more dtypes — this restriction
+# is NVIDIA-specific and should be relaxed per-backend when the framework allows.
```
Comment (tengqm) [inline benchmark/test_special_bessel_j1.py]: okay, then please parametrize the `dtype` parameter, let it be fp32 for CUDA and FLOAT_DTYPES for other platforms.

Comment (XDYuanzhuLee) [reply benchmark/test_special_bessel_j1.py]: Parametrized the `dtype` per backend as requested: `[torch.float32]` for CUDA (matching the CUDA reference constraint) and `consts.FLOAT_DTYPES` for all other backends.

**Comment by bin913** [comment]
@XDYuanzhuLee Please fix failing check

**Review by tengqm — DISMISSED** [review]

**Review by Dongxu-H — COMMENTED** [review]
## FlagGems Operator 贡献审查报告

**Operator**: `special_bessel_j1`
**PR**: #3365

### 📊 变更统计
- 新增行数: 98
- 删除行数: 0  
- 变更文件: 6

### ✅ 检查清单

| 检查项 | 状态 |
|--------|------|
| 实现文件 (`src/flag_gems/ops/special_bessel_j1.py`) | ✅ 存在 |
| 测试文件 (`tests/test_special_bessel_j1.py`) | ✅ 存在 |
| Benchmark 文件 | ✅ 存在 |
| 配置注册 (`operators.yaml`) | ✅ 存在 |
| 模块注册 (`__init__.py`) | ✅ 存在 |
| KernelGen 标注 | ✅ 存在 |
| Triton JIT 装饰器 | ✅ 存在 |
| Metax Backend | ⚠️ 未实现 |

### 📝 审查意见

**代码结构**: 实现文件结构正确

**测试覆盖**: 有测试文件

**性能测试**: 有 benchmark

---

*此审查由自动化脚本根据 FlagGems Operator 贡献规范生成。*
*如有问题请联系维护者进行详细审查。*

**Review by Dongxu-H — COMMENTED** [review]
## FlagGems Operator Contribution Review

**Operator**: `special_bessel_j1`  
**PR**: #3365

### 📊 Change Statistics
| Metric | Value |
|--------|-------|
| Additions | 98 |
| Deletions | 0 |
| Files Changed | 6 |

### 📋 22-Point Checklist

| # | Requirement | Status |
|---|-------------|--------|
| 1 | Implementation file path correctness | ✅ Correct: `src/flag_gems/ops/special_bessel_j1.py` |
| 2 | KernelGen attribution in header | ✅ Present |
| 3 | Import order (stdlib → third-party → internal) | ⚠️ Could not verify |
| 4 | Function signature alignment with PyTorch API | ⚠️ Needs manual verification |
| 5 | Type support documentation | ✅ Type handling present |
| 6 | Triton kernel decorators (`@triton.jit`, etc.) | ✅ `@triton.jit` and decorators present |
| 7 | Test file existence | ✅ Present: `tests/test_special_bessel_j1.py` |
| 8 | Test coverage (shapes, dtypes) | ✅ Shape and dtype tests present |
| 9 | Benchmark file existence | ✅ Present: `benchmark/test_special_bessel_j1.py` |
| 10 | Benchmark implementation correctness | ✅ Benchmark implemented |
| 11 | Config registration (`operators.yaml`) | ✅ Registered in `operators.yaml` |
| 12 | Module registration (`__init__.py`) | ✅ Registered in `__init__.py` |
| 13 | Ops module registration (`ops/__init__.py`) | ✅ Registered in `ops/__init__.py` |
| 14 | Logger namespace (`flag_gems.*`) | ⚠️ Missing `flag_gems.` prefix |
| 15 | Error handling | ✅ Error handling present |
| 16 | Numerical stability patterns | ⚠️ Needs domain expert review |
| 17 | Memory efficiency | ⚠️ Needs profiling |
| 18 | Edge case tests | ⚠️ Needs verification |
| 19 | Backend support (Metax) | ❌ Not implemented |
| 20 | Documentation/Docstrings | ⚠️ Needs manual review |
| 21 | Code style consistency | ⚠️ Run linter manually |
| 22 | Integration verification | ⚠️ Verify end-to-end |

### 📈 Summary

| Category | Count |
|----------|-------|
| ✅ Pass | 12 |
| ⚠️ Warning | 9 |
| ❌ Fail | 1 |

**Overall**: ⚠️ Needs Attention

---

*This review was generated by an automated script following the FlagGems Operator Contribution Guidelines.*
*For detailed code review, please consult a domain expert.*

**Review by Dongxu-H — COMMENTED** [review]
## 💡 审查建议

建议:
- ⚠️ 建议在flag_gems/__init__.py注册

**Review by Dongxu-H — COMMENTED** [review]
## 💡 审查建议

建议:
- ⚠️ 建议在flag_gems/__init__.py注册

**Review by Dongxu-H — COMMENTED** [review]
## 💡 审查建议

建议:
- ⚠️ 建议在flag_gems/__init__.py注册

**Comment by Caeruleann** [comment]
Please remove the Co-Authored-By trailer from the commit message.

**Comment by XDYuanzhuLee** [comment]
Correction: the Co-Authored-By trailer was already absent from the HEAD commit (ea5bbd9) when this review was posted. No further changes were needed — the fix predated the review. Note: 3 older commits in the branch history (4d94568, 9c42cb3, 26cda26) still contain Co-Authored-By trailers and will be addressed separately.

**Comment by XDYuanzhuLee** [comment]
Verified: the Co-Authored-By trailer was already removed from the HEAD commit (ea5bbd9). No further changes needed.

**Review by Dongxu-H — APPROVED** [review]
## 代码审查结果

### ✅ 全部检查通过

| 检查项 | 状态 |
|--------|------|
| CI 状态 | ✅ PASS (python-op, code-style, license/cla) |
| 合并冲突 | ✅ 无冲突 (MERGEABLE) |
| 配置排序 (operators.yaml) | ✅ `special_bessel_j1` 在 `sparse_mla_fwd_interface` 后，`special_chebyshev_polynomial_v` 前 |
| 配置排序 (__init__.py) | ✅ `special_bessel_j1` 在 `sort.stable` 后，`special_chebyshev_polynomial_v` 前 |
| 配置排序 (ops/__init__.py) | ✅ 导入和 `__all__` 顺序正确 |
| KernelGen 标注 | ✅ 头部存在 |
| Logger 命名 | ✅ `logging.getLogger(__name__)` → `flag_gems.ops.special_bessel_j1` |
| Triton JIT | ✅ `@pointwise_dynamic` + `@triton.jit` |
| 测试覆盖 | ✅ `POINTWISE_SHAPES` + `FLOAT_DTYPES` |
| Benchmark 文件 | ✅ 存在（含 CUDA dtype 处理） |
| flag_gems/__init__.py 注册 | ✅ 已注册 `special_bessel_j1` |
| Dtype 支持 | ✅ FLOAT_DTYPES |

---

### 📝 实现亮点

1. **简洁正确的实现**：
   ```python
   @triton.jit
   def j1_func(x):
       return tl_extra_shim.j1(x.to(tl.float32))
   ```
   使用 `tl_extra_shim.j1` 调用 Triton 的 Bessel J1 函数。

2. **高精度处理**：
   - 内部转换为 float32 进行计算（Bessel 函数需要高精度）
   - 正确处理 fp16/bf16 dtype

3. **CUDA dtype 适配**：
   - 测试中正确处理 CUDA 的 dtype 限制
   - Benchmark 中区分 CUDA 和其他 backend 的 dtype 支持

4. **多 backend 测试**：
   - Nvidia: ✅ PASS
   - Hygon: ✅ PASS
   - 性能: 6.341x 平均加速

---

### 📋 总结

PR 实现正确，配置完整，测试覆盖充分，性能显著提升。推荐合并。

**Review by Dongxu-H — DISMISSED** [review]

File: benchmark/test_special_bessel_j1.py (hunk 1 above)
Comment (bin913) [reply benchmark/test_special_bessel_j1.py]: The test fail for half dtype fail in non-NVIDIA device. This is conflict with the comment

Comment (XDYuanzhuLee) [reply benchmark/test_special_bessel_j1.py]: Fixed. The benchmark now uses float32 only on all backends (same pattern as special_hermite_polynomial_h and special_chebyshev_polynomial_v), and the test no longer has a device-specific check -- it always computes the reference in float32 for half/bfloat16 inputs and casts back. Both changes pushed to the branch.

**Review by Caeruleann — DISMISSED** [review]

**Comment by bin913** [comment]
@XDYuanzhuLee Please resolve conflict

**Comment by Caeruleann** [comment]
Please resolve conflicts.

**Review by bin913 — APPROVED** [review]
